dart/lib/src/http_client/sentry_http_client.dart

```diff
@@ -0,0 +1,99 @@
+/// which records requests as breadcrumbs.
+///
+/// Remarks:
+/// If this client is used as a wrapper, a call to close does not close the
+/// given client.
+///
+/// The `SentryHttpClient` can be used as a standalone client like this:
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+///
+/// var client = SentryHttpClient();
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+///
+/// The `SentryHttpClient` can also be used as a wrapper for your own
+/// HTTP [Client](https://pub.dev/documentation/http/latest/http/Client-class.html):
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+/// import 'package:http/http.dart' as http;
+///
+/// final myClient = http.Client();
+///
+/// var client = SentryHttpClient(client: myClient);
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+//
+// Possible enhancement:
+// Track the time the HTTP request took.
+// For example with Darts Stopwatch:
+// https://api.dart.dev/stable/2.10.4/dart-core/Stopwatch-class.html
+class SentryHttpClient extends BaseClient {
+  SentryHttpClient({Client client, Hub hub}) {
+    _hub = hub ?? HubAdapter();
+    if (client != null) {
+      _client = client;
+      _userdefinedClient = true;
+    } else {
+      _client == Client();
+    }
+  }
+
+  Client _client;
+  Hub _hub;
+  bool _userdefinedClient = false;
+
+  @override
+  Future<StreamedResponse> send(BaseRequest request) async {
+    final response = await _client.send(request);
+    // See https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/
+    // for how the breadcrumb should look like
+    _hub.addBreadcrumb(
+      Breadcrumb(
+        // TODO: should this always be `http`?
+        // Otherwise use http or https depending on the request.
+        type: request.url.scheme,
+        data: {
+          'url': request.url.toString(),
+          'method': request.method,
+          'status_code': response.statusCode,
+          // reason is optional, therefor only add it in case it is not null
+          if (response.reasonPhrase != null) 'reason': response.reasonPhrase,
+        },
+      ),
+    );
+    return response;
+  }
+
+  @override
+  void close() {
+    // In case the user did not specify a client himself
+    // we should close the client we initialized.
+    //
+    // If the user did specifiy a client himself,
+    // the user is in charge of closing his own client.
+    // Thus we don't close it here.
+    if (!_userdefinedClient) {
+      // TODO: how to test this?
+      // There is no property on a Client which indicates its status.
+      _client.close();
```

Comment:
true, I'd just validate if the method was or not called, using a mock maybe, or you create a fake that sets a flag so you know if its called or not.

Reply:
Yep, but I only want to call it in case this client was initialized by the constructor so `SentryHttpClient` closes its own client but not the one given by the user. So mocking cannot be used because this code should not be called. `SentryHttpClient` should not interfere with the lifespan of the users http client.

Reply:
Out of date because of https://github.com/getsentry/sentry-dart/pull/226#discussion_r536984785

Reply:
but that's what I meant, using `verify` and `verifyNever`

> if the method was or not called

Reply:
You can only verify if you're using a mock. But in this case I can't use a mock, so no `verify` :(

But well, at least https://github.com/getsentry/sentry-dart/pull/226#discussion_r536984785 saves us from this headache.

Reply:
yeah, lets spare the headache.

but fwiw, your test called `close does not get called for user defined client` does exactly what I've suggested and you do use a Mock in there.

Reply:
but yeah, you'd like to test the usecase that the Client is created internally in the wrapper, true, well, solved by not doing it :)